File: app/app/src/main/java/williamhester/me/breadit2/ui/text/VotableMovementMethod.java
package williamhester.me.breadit2.ui.text;

import android.support.annotation.NonNull;
import android.text.Spannable;
import android.text.method.LinkMovementMethod;
import android.view.MotionEvent;
import android.widget.TextView;

import williamhester.me.breadit2.models.Votable;
import williamhester.me.breadit2.ui.VotableCallbacks;

/**
 * A {@link LinkMovementMethod} that propogates the click to a {@link VotableCallbacks} if a link is
 * not clicked.
 */
public class VotableMovementMethod extends LinkMovementMethod {

  private final VotableCallbacks votableCallbacks;
  private Votable votable;

  public VotableMovementMethod(VotableCallbacks votableCallbacks) {
    this.votableCallbacks = votableCallbacks;
  }

  public void setVotable(Votable votable) {
    this.votable = votable;
  }

  @Override
  public boolean onTouchEvent(@NonNull TextView widget, @NonNull Spannable buffer,
                              @NonNull MotionEvent event) {
    if (event.getAction() == MotionEvent.ACTION_UP) {
      if (!super.onTouchEvent(widget, buffer, event)) {
        votableCallbacks.onVotableClicked(votable);
        return false;
      } else {
        return true;
      }
    }
    return super.onTouchEvent(widget, buffer, event);
  }
}

File: app/app/src/main/java/me/williamhester/reddit/ui/text/VotableMovementMethod.kt
package me.williamhester.reddit.ui.text

import android.text.Spannable
import android.text.method.LinkMovementMethod
import android.view.MotionEvent
import android.widget.TextView

import me.williamhester.reddit.models.Votable
import me.williamhester.reddit.ui.VotableCallbacks

/**
 * A [LinkMovementMethod] that propogates the click to a [VotableCallbacks] if a link is not
 * clicked.
 */
class VotableMovementMethod(private val votableCallbacks: VotableCallbacks) : LinkMovementMethod() {
  private var votable: Votable? = null

  fun setVotable(votable: Votable) {
    this.votable = votable
  }

  override fun onTouchEvent(widget: TextView, buffer: Spannable, event: MotionEvent): Boolean {
    if (event.action == MotionEvent.ACTION_UP) {
      if (!super.onTouchEvent(widget, buffer, event)) {
        votableCallbacks.onVotableClicked(votable!!)
        return false
      } else {
        return true
      }
    }
    return super.onTouchEvent(widget, buffer, event)
  }
}
